indy_node/test/catchup/test_state_recovering.py

```diff
@@ -0,0 +1,133 @@
+    for n in nodeSet[0].poolManager.nodeIds:
+        schedule[n] = datetime.isoformat(startAt)
+        startAt = startAt + timedelta(seconds=acceptableDiff + 3)
+
+    return dict(name='upgrade', version="10000.10.10",
+                action=START, schedule=schedule, timeout=1,
+                package=None,
+                sha256='db34a72a90d026dae49c3b3f0436c8d3963476c77468ad955845a1ccf7b03f55')
+
+
+def send_node_upgrades(nodes, version, looper, count=None):
+    if count is None:
+        count = len(nodes)
+    last_ordered = nodes[0].master_last_ordered_3PC[1]
+    for node in nodes[:count]:
+        op = {
+            TXN_TYPE: NODE_UPGRADE,
+            DATA: {
+                ACTION: COMPLETE,
+                VERSION: version
+            }
+        }
+        op[f.SIG.nm] = node.wallet.signMsg(op[DATA])
+
+        request = node.wallet.signRequest(
+            Request(operation=op, protocolVersion=CURRENT_PROTOCOL_VERSION))
+
+        node.startedProcessingReq(request.key, node.nodestack.name)
+        node.send(request)
+    looper.run(eventually(lambda: assertEquality(nodes[0].master_last_ordered_3PC[1],
+                                                 last_ordered + count)))
+
+
+def send_auth_rule(looper,
+                   sdk_pool_handle,
+                   sdk_wallet_trustee,
+                   original_action):
+    sdk_send_and_check_auth_rule_request(looper,
+                                         sdk_pool_handle,
+                                         sdk_wallet_trustee,
+                                         auth_action=ADD_PREFIX, auth_type=original_action.txn_type,
+                                         field=original_action.field, new_value=original_action.value,
+                                         old_value=None, constraint=AuthConstraint(role="*",
+                                                                                   sig_count=1).as_dict)
+
+
+def test_state_recovering_for_auth_rule(nodeSet, looper, sdk_pool_handle, sdk_wallet_steward,
+                                        sdk_wallet_trustee,
+                                        tdir, tconf,
+                                        allPluginsPath,
+                                        valid_upgrade,
+                                        auth_keys,
+                                        monkeypatch):
+    version1 = "1.1.58"
+    version2 = "1.1.88"
+
+    # send POOL_UPGRADE to write in a ledger
+    sdk_ensure_upgrade_sent(looper, sdk_pool_handle, sdk_wallet_trustee,
+                            valid_upgrade)
+    send_node_upgrades(nodeSet, version1, looper, count=1)
+    send_auth_rule(looper,
+                   sdk_pool_handle,
+                   sdk_wallet_trustee,
+                   auth_keys[0])
+    send_node_upgrades(nodeSet, version1, looper)
+    for n in nodeSet:
+        handler = n.write_manager.request_handlers.get(AUTH_RULE)[0]
+        monkeypatch.setattr(handler, '_update_auth_constraint',
```

Comment:
Why do we do monkey patching here? This doesn't check that we chose valid state update strategy, does it?

Reply:
We patch ```update_state``` there to write AUTH_RULE txns in the old format for ordinary nodes.